### ISSUE TITLE
Question: arrayContaining Matcher and the dynamic JSON keys.

### ISSUE THREAD

**Issue by StasMarynych** [surpher/PactSwift#108]
# ❔ Question

Are there any possibility to define the expected response in case of dynamic JSON keys (set of the different JSON objects)? For example something like [arrayContaining](https://docs.pact.io/implementation_guides/javascript/docs/matching) in _pact-js_.

# 💬 Context

The main problem is that the dynamic JSON objects are located in the array, so there can be more than 1 different objects in the array. There is a predefined set of the possible objects, so if there are no any limitations I would consider implementing same matcher as in _pact-js_ version. (I'd love to help with that or v4 specification implementation.)

**JSON Sample**

```
{
  "items": [
    {
      "id": "some-id1",
      "config": { 
        "foo": "true"
      }
    },
    {
      "id": "some-id2",
      "config": {
        "bar": { ... }
      }
    },
    {
      "id": "some-id3",
      "config": { 
        "baz": 100 
       }
    }
  ]
}
```

# 🤝 Relationships
Looks like in v4 specification combine operators could be used, so that will solve the issue as well.

- Other Related Issues: #107

**Comment by StasMarynych** [comment]
Just noticed that [arrayContaining](https://docs.pact.io/implementation_guides/javascript/docs/matching) method is implemented in v3, are there any chances that we could easily add that matcher without v4 ?

**Comment by StasMarynych** [comment]
I was able to implement test matcher in my fork but I am not sure whether contact format is valid or not. https://github.com/StasMarynych/PactSwift/tree/feat/implement-containing-like-matcher

**Comment by StasMarynych** [comment]
I was able to implement some sort of `arrayContains` matcher. Response generation is ok, also i tried to add matching rule by doing this:

```
processedMatcherValue.rules[${node_with_array_contains}] = AnyEncodable([
	"matchers": [
		"match": AnyEncodable("arrayContains"),
		"variants": AnyEncodable(${all_processed_variants})
	]
])
```
but it wont appear in output pact contact

**Comment by StasMarynych** [comment]
Finally, got the example of `arrayContains` rule using `pact_mock_server_cli`. Will refactor matcher processing to match the appropriate format

```
"$.baz": {
              "combine": "AND",
              "matchers": [
                {
                  "match": "arrayContains",
                  "variants": [
                    {
                      "index": 0,
                      "rules": {
                        "": {
                          "combine": "AND",
                          "matchers": [
                            {
                              "match": "equality"
                            }
                          ]
                        }
                      }
                    },
                    {
                      "index": 0,
                      "rules": {
                        "": {
                          "combine": "AND",
                          "matchers": [
                            {
                              "match": "equality"
                            }
                          ]
                        }
                      }
                    }
                  ]
                }
              ]
            }
```

**Comment by surpher** [comment]
Hey @StasMarynych 
Sorry for late reply. I'm unfortunately quite far away from GitHub lately. I saw your closed PR, what happened there?

I've pulled the fork that refactored and massively rebased PactSwift code to take advantage of `libpact_ffi` more. 
Changes are on this branch https://github.com/surpher/PactSwift/tree/feat/v2.0.0 and it has full on support for modern swift concurrency.

**Comment by StasMarynych** [comment]
Hey @surpher 

That mr was opened by the accident actually. Firstly, I wanted to be sure that current version of libpact_ffi that is in PactSwiftMockServer package support that new type of matcher at all. Now, I see that everything is ok. So, if there is no `arrayContains` matcher in https://github.com/surpher/PactSwift/tree/feat/v2.0.0 that would be better to create the MR based on this branch?

**Comment by surpher** [comment]
Yes sure. There might be a caveat though where v2 is full on concurrency focused. Changes the set up and structure of pact tests. 
Make sure first v2 fits your needs.
